File: Dsa-day35.py
class Solution:
    def numSubmat(self, mat):
        m, n = len(mat), len(mat[0])
        heights = [0] * n
        ans = 0
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 1:
                    heights[j] += 1
                else:
                    heights[j] = 0
            for j in range(n):
                h = heights[j]
                k = j
                while k >= 0 and h > 0:
                    h = min(h, heights[k])
                    ans += h
                    k -= 1
        return ans

File: Dsa-day35.java
class Solution {
    public int numSubmat(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        int[] heights = new int[n];
        int ans = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 1) heights[j]++;
                else heights[j] = 0;
            }
            for (int j = 0; j < n; j++) {
                int h = heights[j];
                for (int k = j; k >= 0 && h > 0; k--) {
                    h = Math.min(h, heights[k]);
                    ans += h;
                }
            }
        }
        return ans;
    }
}
